File: src/main/java/pia/logic/VideoWriter.java
package pia.logic;

import pia.database.Database;
import pia.database.model.archive.Video;
import pia.exceptions.CreateHashException;
import pia.filesystem.BufferedFileWithMetaData;
import pia.filesystem.FileSystemHelper;
import pia.tools.FileHash;
import pia.tools.Logger;

import java.io.File;
import java.io.IOException;
import java.util.Optional;

public class VideoWriter {
    private final Logger logger = new Logger(VideoWriter.class);

    public void addVideo(BufferedFileWithMetaData bufferedFile, String fileName) throws IOException, CreateHashException {
        Optional<String> hash = FileHash.createHash(bufferedFile);
        if(hash.isPresent()) {
            Video video = new Video();
            FileSystemHelper fileSystemHelper = new FileSystemHelper();
            String fileOnDisk = video.getId().toString() + "." + fileSystemHelper.getFileExtension(fileName);
            File file = fileSystemHelper.writeFileToDisk(bufferedFile, fileOnDisk);
            if (file.exists()) {
                video.setSha256Hash(hash.get());
                video.setOriginalFileName(fileName);
                video.setPathToFileOnDisk(file.getAbsolutePath());
                video.setCreationTime(bufferedFile.getCreationDate());
                Database.getConnection().insertObject(video);
            } else {
                logger.error("error writing file to disk");
            }
        } else {
            throw new CreateHashException();
        }
    }

    public boolean deleteVideo(Video video) {
        boolean deleted = false;
        String videoFile = video.getPathToFileOnDisk();
        FileSystemHelper fileSystemHelper = new FileSystemHelper();
        boolean deleteFromDb = true;
        if (fileSystemHelper.fileExists(videoFile)) {
            if (!fileSystemHelper.deleteFileFromDisk(videoFile)) {
                deleteFromDb = false;
                logger.error("could not delete videofile from disk %s, db-entry will not be deleted", videoFile);
            }
        } else {
            logger.warn("deleting video %s, file %s does not exist", video.getId(), video.getPathToFileOnDisk());
        }
        if (deleteFromDb) {
            if (Database.getConnection().deleteObject(video)) {
                deleted = true;
            } else {
                logger.error("could not delete the videoobject %s from database", video.getId());
            }
        }
        return deleted;
    }
}

File: src/src/main/java/pia/logic/VideoWriter.kt
package pia.logic

import mu.KotlinLogging
import pia.database.Database
import pia.database.model.archive.Video
import pia.exceptions.CreateHashException
import pia.filesystem.FileSystemHelper
import pia.filesystem.MediaType
import pia.logic.tools.VideoInfo
import pia.tools.toJodaDateTime
import java.io.IOException

class VideoWriter {
    private val logger = KotlinLogging.logger {  }

    fun addVideo(sourceFilePath : String, videoInfo: VideoInfo, originalFileName : String) : Video? {
        var video : Video? = null
        video = Video.new {
            val fileSystemHelper = FileSystemHelper()
            val file = fileSystemHelper.moveFileToArchive(sourceFilePath, originalFileName, videoInfo.getCreationDate().year, videoInfo.getCreationDate().month, MediaType.Video)
            if (file.exists()) {
                this.originalFileName = originalFileName
                pathToFileOnDisk = file.absolutePath
                creationTime = videoInfo.getCreationDate().toJodaDateTime()
            } else {
                logger.error("error writing file to disk")
            }
        }
        return video
    }

    fun deleteVideo(video: Video, force : Boolean): Boolean {
        var deleted = false
        Database.connection.transactional {
            val videoFile = video.pathToFileOnDisk
            val fileSystemHelper = FileSystemHelper()
            var deleteFromDb = true
            if (fileSystemHelper.fileExists(videoFile.orEmpty())) {
                if (!fileSystemHelper.deleteFileFromDisk(videoFile.orEmpty())) {
                    deleteFromDb = force
                    logger.error("could not delete videofile from disk $videoFile, delete db-entry: $force")
                }
            } else {
                logger.warn("deleting video {}, file {} does not exist", video.entityId, video.pathToFileOnDisk)
            }
            if (deleteFromDb) {
                video.delete()
                deleted = true
            }
            it.commit()
        }
        return deleted
    }
}